GitUI/UserControls/NativeTreeView.cs

```diff
@@ -8,4 +12,100 @@ protected override void CreateHandle()
             NativeMethods.SetWindowTheme(Handle, "explorer", null);
         }
     }
+
+    // Add double-click related behaviour to NativeTreeView
+    public class NativeTreeViewDoubleClickDecorator
+    {
+        private NativeTreeView _treeView;
+
+        private DateTime _lastMouseDown = DateTime.MinValue;
+        private int _mouseClicks = 0;
+
+        private DateTime _lastDoubleClick = DateTime.MinValue;
+
+        // Invoked just before an inner tree node is expanded/collapsed. Set CancelEventHandler.Cancel to true
+        // to cancel the expand/collapse of the current node.
+        public event CancelEventHandler BeforeDoubleClickExpandCollapse;
+
+        public NativeTreeViewDoubleClickDecorator(NativeTreeView treeView)
+        {
+            _treeView = treeView;
+            _treeView.Disposed += OnDisposed;
+            _treeView.BeforeCollapse += OnBeforeCollapse;
+            _treeView.BeforeExpand += OnBeforeExpand;
+            _treeView.MouseDown += OnMouseDown;
+        }
+
+        private void OnDisposed(object sender, EventArgs e)
+        {
+            _treeView.BeforeCollapse -= OnBeforeCollapse;
+            _treeView.BeforeExpand -= OnBeforeExpand;
+            _treeView.MouseDown -= OnMouseDown;
+            _treeView = null;
+        }
+
+        private void OnBeforeCollapse(object sender, TreeViewCancelEventArgs e)
+        {
+            HandleBeforeDoubleClickExpandCollapse(sender, e);
+        }
+
+        private void OnBeforeExpand(object sender, TreeViewCancelEventArgs e)
+        {
+            HandleBeforeDoubleClickExpandCollapse(sender, e);
+        }
```

Comment:
There is no real need in `OnBeforeCollapse` and `OnBeforeExpand`, just wire up directly to `HandleBeforeDoubleClickExpandCollapse`:
```
_treeView.BeforeCollapse += HandleBeforeDoubleClickExpandCollapse;
_treeView.BeforeExpand += HandleBeforeDoubleClickExpandCollapse;
```